File: lib/src/services/conjugation.dart
library app.src.services.conjugation;

import 'package:angular2/angular2.dart' show Injectable;
import 'verbs.dart';

bool _isFirstPerson(Person p) => p == Person.First;
bool _isSecondPerson(Person p) => p == Person.Second;
bool _isThirdPerson(Person p) => p == Person.Third;
bool _isSingular(Plurality p) => p == Plurality.Singular;
bool _isPlural(Plurality p) => p == Plurality.Plural;

enum Tense { Simple, Progressive, Perfect, PerfectProgressive }

enum Time { Present, Past, Future, Conditional }

enum Person { First, Second, Third }

enum Plurality { Singular, Plural }

_SimpleTenseOfBe(Time time, Person person, Plurality plurality) {
  bool isFirst = _isFirstPerson(person);
  bool isSecond = _isSecondPerson(person);
  bool isThird = _isThirdPerson(person);
  bool isSingular = _isSingular(plurality);
  bool isPlural = _isPlural(plurality);

  if (time == Time.Present) if (isFirst && isSingular) return "am";
  else return (isFirst || isSecond || isThird && isPlural) ? "are" : "is";
  else if (time == Time.Past) return (isFirst && isSingular ||
      isThird && isSingular) ? "was" : "were";
  else if (time == Time.Future) return "will be";
  else if (time == Time.Conditional) return "would be";
}

String _SimpleTenseOfHave(Time time, Person person, Plurality plurality) {
  if (time == Time.Present) return (_isThirdPerson(person) &&
      _isSingular(plurality)) ? "has" : "have";
  else if (time == Time.Past) return "had";
  else if (time == Time.Future) return "will have";
  else return "would have";
}

abstract class Conjugation {
  get infinitive;

  get(Tense tense, Time time, Person person, Plurality plurality);

  getConjugationTable(Tense tense);
}

abstract class _DefaultConjugationTable implements Conjugation {
  List _structure = [
    ["I", Person.First, Plurality.Singular],
    ["You", Person.Second, Plurality.Singular],
    ["He", Person.Third, Plurality.Singular],
    ["She", Person.Third, Plurality.Singular],
    ["It", Person.Third, Plurality.Singular],
    ["We", Person.First, Plurality.Plural],
    ["They", Person.Third, Plurality.Plural]
  ];

  getConjugationTable(Tense tense) {
    List items = [];
    _structure.forEach((List e) {
      List item = [];
      Time.values.forEach((Time time) {
        String c = this.get(tense, time, e.elementAt(1), e.elementAt(2));
        item.add("${e.elementAt(0)} ${c}");
      });
      items.add(item);
    });
    return items;
  }
}

class _DefaultConjugation extends Conjugation with _DefaultConjugationTable {
  Map<String, String> _data;

  _DefaultConjugation(this._data);

  get infinitive => _data['1'];
  get participle => _data['4'];
  get progressive => _data['3'];
  get past => _data['2'];
  get thirdPerson => _data['5'];

  String get(Tense tense, Time time, Person person, Plurality plurality) {
    if (tense == Tense.Simple) return _getSimpleTenseConjugation(
        tense, time, person, plurality);
    else if (tense ==
        Tense.Progressive) return "${_SimpleTenseOfBe(time,person,plurality)} ${this.progressive}";
    else if (tense ==
        Tense.Perfect) return "${_SimpleTenseOfHave(time,person,plurality)} ${this.participle}";
    else return "${_SimpleTenseOfHave(time,person,plurality)} been ${this.progressive}";
  }

  _getSimpleTenseConjugation(
      Tense tense, Time time, Person person, Plurality plurality) {
    if (time == Time.Present) {
      return (_isThirdPerson(person) && _isSingular(plurality))
          ? this.thirdPerson
          : this.infinitive;
    } else if (time == Time.Past) return past;
    else if (time == Time.Conditional) return "would ${this.infinitive}";
    else if (time == Time.Future) return "will ${this.infinitive}";
  }
}

class _BeConjugation extends Conjugation with _DefaultConjugationTable {
  Map<String, String> _data;

  _BeConjugation(this._data);

  get infinitive => _data['1'];
  get participle => _data['2'];
  get progressive => _data['3'];

  get(Tense tense, Time time, Person person, Plurality plurality) {
    if (tense == Tense.Simple) return _SimpleTenseOfBe(time, person, plurality);
    else if (tense ==
        Tense.Progressive) return "${_SimpleTenseOfBe(time,person,plurality)} ${this.progressive}";
    else if (tense ==
        Tense.Perfect) return "${_SimpleTenseOfHave(time,person,plurality)} ${this.participle}";
    else return "${_SimpleTenseOfHave(time,person,plurality)} been ${this.progressive}";
  }
}

@Injectable()
class ConjugationService {

  _normalize(String search){
    search = search.trim().toLowerCase();
    if(search.startsWith('to ') && search.length > 3)
      search = search.substring(3,search.length);
    return search;
  }

  Conjugation find(String verb) {
    verb = _normalize(verb);
    Map<String, String> data = this._findByInfinitive(verb.toLowerCase());
    if (data['1'] == 'be') {
      return new _BeConjugation(data);
    }
    return new _DefaultConjugation(data);
  }

  Map<String, String> _findByInfinitive(String verb) {
    return verbs.singleWhere((data) {
      return data['1'] == verb;
    });
  }
}

File: libs/conjugation/conjugation.ts
import { Person, Plurality, Tense, Time, VerbData } from './types'

const structure: [string, Person, Plurality][] = [
  ['I', Person.First, Plurality.Singular],
  ['You', Person.Second, Plurality.Singular],
  ['He', Person.Third, Plurality.Singular],
  ['She', Person.Third, Plurality.Singular],
  ['It', Person.Third, Plurality.Singular],
  ['We', Person.First, Plurality.Plural],
  ['They', Person.Third, Plurality.Plural],
]

export function getSimpleTenseOfBe(
  time: Time,
  person: Person,
  plurality: Plurality
) {
  const isFirst = person === Person.First
  const isSecond = person === Person.Second
  const isThird = person === Person.Third
  const isSingular = plurality === Plurality.Singular
  const isPlural = plurality === Plurality.Plural

  if (time == Time.Present)
    if (isFirst && isSingular) return 'am'
    else return isFirst || isSecond || (isThird && isPlural) ? 'are' : 'is'
  else if (time == Time.Past)
    return (isFirst && isSingular) || (isThird && isSingular) ? 'was' : 'were'
  else if (time == Time.Future) return 'will be'
  else return 'would be'
}

export function getSimpleTenseOfHave(
  time: Time,
  person: Person,
  plurality: Plurality
) {
  const isThird = person === Person.Third
  const isSingular = plurality === Plurality.Singular

  if (time == Time.Present) return isThird && isSingular ? 'has' : 'have'
  else if (time == Time.Past) return 'had'
  else if (time == Time.Future) return 'will have'
  else return 'would have'
}

export abstract class Conjugation {
  abstract readonly data: VerbData
  abstract get infinitive(): string
  public abstract get(
    tense: Tense,
    time: Time,
    person: Person,
    plurality: Plurality
  ): string

  private getRows(tense: Tense, time: Time) {
    return structure.map(
      ([subject, person, plurality]) =>
        `${subject} ${this.get(tense, time, person, plurality)}`
    )
  }

  getTable(tense: Tense) {
    return {
      [Time.Present]: this.getRows(tense, Time.Present),
      [Time.Past]: this.getRows(tense, Time.Past),
      [Time.Future]: this.getRows(tense, Time.Future),
      [Time.Conditional]: this.getRows(tense, Time.Conditional),
    }
  }
}

class DefaultConjugation extends Conjugation {
  constructor(readonly data: VerbData) {
    super()
  }

  get infinitive() {
    return this.data['1']
  }

  get progressive() {
    return this.data['3']
  }

  public get(tense: Tense, time: Time, person: Person, plurality: Plurality) {
    if (tense == Tense.Simple)
      return this.getSimpleTenseConjugation(time, person, plurality)
    else if (tense == Tense.Progressive)
      return `${getSimpleTenseOfBe(time, person, plurality)} ${
        this.progressive
      }`
    else if (tense == Tense.Perfect)
      return `${getSimpleTenseOfHave(time, person, plurality)} ${
        this.data['4']
      }`
    else
      return `${getSimpleTenseOfHave(time, person, plurality)} been ${
        this.progressive
      }`
  }

  private getSimpleTenseConjugation(
    time: Time,
    person: Person,
    plurality: Plurality
  ): string {
    const isThird = person === Person.Third
    const isSingular = plurality === Plurality.Singular

    if (time == Time.Present) {
      return isThird && isSingular ? this.data['5'] : this.infinitive
    } else if (time == Time.Past) return this.data['2'] as string
    else if (time == Time.Conditional) return `would ${this.infinitive}`
    else return `will ${this.infinitive}`
  }
}

class BeConjugation extends DefaultConjugation {
  public get(tense: Tense, time: Time, person: Person, plurality: Plurality) {
    if (tense === Tense.Simple) {
      return getSimpleTenseOfBe(time, person, plurality)
    } else if (tense === Tense.Progressive) {
      return `${getSimpleTenseOfBe(time, person, plurality)} ${
        this.progressive
      }`
    } else if (tense == Tense.Perfect)
      return `${getSimpleTenseOfHave(time, person, plurality)} ${
        this.data['2']
      }`
    else
      return `${getSimpleTenseOfHave(time, person, plurality)} been ${
        this.progressive
      }`
  }
}

export function getConjugation(data: VerbData): Conjugation {
  const isBe = data['1'] === 'be'
  return isBe ? new BeConjugation(data) : new DefaultConjugation(data)
}
